File: examples/winstt-electron/server/scripts/crap/analyzer.py
"""Orchestrator. Mirrors `frontend/scripts/crap/analyzer.ts`.

Walks every `src/**/*.py` source file, derives per-function complexity +
coverage, then combines them into a CRAP score.

Exclusion model (the Python analog of the frontend's biome-disabled +
crap.ignore.json merge): the project's `[tool.coverage.run].omit` globs in
``pyproject.toml`` already declare what is outside the coverage bar
(infrastructure adapters, bootstrap, client, the WebSocket server, the
terminal shim). We honor that same list so CRAP scope == coverage scope.
An optional ``server/crap.ignore.json`` ({"ignore": [...]}) can add more.
"""

from __future__ import annotations

import json
import re
import tomllib
from dataclasses import dataclass
from pathlib import Path

from .coverage_parser import CoverageMap
from .method_parser import parse_functions
from .score import calculate_crap_score


@dataclass
class FunctionMetric:
    file: str
    name: str
    start_line: int
    end_line: int
    complexity: int
    coverage: float | None  # fraction in [0, 1]
    crap: float | None


def read_coverage_omit(project_root: Path) -> list[str]:
    """Return `[tool.coverage.run].omit` patterns from pyproject.toml."""
    pyproject = project_root / "pyproject.toml"
    if not pyproject.exists():
        return []
    data = tomllib.loads(pyproject.read_text(encoding="utf-8"))
    omit = data.get("tool", {}).get("coverage", {}).get("run", {}).get("omit", [])
    return [str(p).replace("\\", "/") for p in omit]


def read_crap_ignore(project_root: Path) -> list[str]:
    """Read optional `server/crap.ignore.json` -> its `ignore` array.

    Decouples CRAP scope from coverage config, exactly like the frontend's
    `crap.ignore.json` decouples it from biome. Empty if the file is absent.
    """
    ignore_path = project_root / "crap.ignore.json"
    if not ignore_path.exists():
        return []
    config = json.loads(ignore_path.read_text(encoding="utf-8"))
    return [str(p).replace("\\", "/") for p in config.get("ignore", [])]


def _glob_to_regexp(pattern: str) -> re.Pattern[str]:
    out = ""
    i = 0
    while i < len(pattern):
        ch = pattern[i]
        if ch == "*":
            if i + 1 < len(pattern) and pattern[i + 1] == "*":
                out += ".*"
                i += 1
            else:
                # coverage.py treats `*` as matching path separators too, so a
                # single `*` here is greedy across slashes (matches its omit
                # semantics, e.g. `src/stt_server/*` -> nested files).
                out += ".*"
        elif ch in r".+?^${}()|[]\\":
            out += "\\" + ch
        else:
            out += ch
        i += 1
    return re.compile(f"^{out}$")


def _path_matches_any(path: str, patterns: list[str]) -> bool:
    for p in patterns:
        if p == path:
            return True
        if p.endswith("/*") and "*" not in p[:-2]:
            if path.startswith(p[:-2] + "/"):
                return True
            continue
        if "*" in p and _glob_to_regexp(p).match(path):
            return True
    return False


def _collect_crap_exclude_lines(text: str) -> set[int]:
    """1-indexed line numbers whose nearest preceding lines carry a
    ``@crap-exclude`` marker (parity with the TS port's annotation scan)."""
    out: set[int] = set()
    for i, line in enumerate(text.splitlines()):
        if "@crap-exclude" in line:
            for j in range(7):
                out.add(i + 1 + j)
    return out


def _function_coverage_fraction(cov_executed: set[int], cov_missing: set[int], start: int, end: int) -> float | None:
    """hit lines / executable lines within [start, end]. Lines outside the
    coverage record set (comments, signatures, type-only code) are treated
    as non-executable. None when the span has no executable lines."""
    executable = 0
    hit = 0
    for ln in range(start, end + 1):
        in_exec = ln in cov_executed
        in_miss = ln in cov_missing
        if not (in_exec or in_miss):
            continue
        executable += 1
        if in_exec:
            hit += 1
    if executable == 0:
        return None
    return hit / executable


def analyze(
    project_root: Path,
    coverage: CoverageMap,
    *,
    roots: list[str],
    exclude_paths: list[str],
) -> list[FunctionMetric]:
    out: list[FunctionMetric] = []
    for root in roots:
        base = project_root / root
        if not base.exists():
            continue
        for path in sorted(base.rglob("*.py")):
            rel = path.relative_to(project_root).as_posix()
            if _path_matches_any(rel, exclude_paths):
                continue
            text = path.read_text(encoding="utf-8")
            excluded_lines = _collect_crap_exclude_lines(text)
            file_cov = coverage.get(rel)
            for fn in parse_functions(rel, text):
                if fn.start_line in excluded_lines:
                    continue
                cov: float | None
                if file_cov is None:
                    cov = None
                else:
                    cov = _function_coverage_fraction(file_cov.executed, file_cov.missing, fn.start_line, fn.end_line)
                out.append(
                    FunctionMetric(
                        file=rel,
                        name=fn.name,
                        start_line=fn.start_line,
                        end_line=fn.end_line,
                        complexity=fn.complexity,
                        coverage=cov,
                        crap=calculate_crap_score(fn.complexity, cov),
                    )
                )
    return out

File: examples/winstt-electron/frontend/scripts/crap/analyzer.ts
/**
 * Orchestrator. Mirrors `examples/crap4java/src/crap4java/CrapAnalyzer.java`.
 * Walks every source file, derives per-function complexity + coverage, then
 * combines them into a CRAP score.
 */

import { resolve } from "node:path";
import { Glob } from "bun";
import type { LcovCoverage } from "./lcov-parser";
import { parseFunctions } from "./method-parser";
import { calculateCrapScore } from "./score";

export interface FunctionMetric {
	complexity: number;
	coverage: number | null; // fraction in [0, 1]
	crap: number | null;
	endLine: number;
	file: string;
	name: string;
	startLine: number;
}

export interface AnalyzeOptions {
	exclude: RegExp; // e.g. /\.(test|spec|d)\.tsx?$/
	/** Extra path patterns (forward-slashed) to skip — used for biome-disabled
	 * files and other principled exclusions. */
	excludePaths?: string[];
	include: RegExp; // e.g. /\.tsx?$/
	roots: string[]; // e.g. ["src", "electron"]
}

/**
 * Parse biome.jsonc and return the list of file paths that have linter.enabled = false.
 * These files are explicitly accepted as outside the quality bar (third-party
 * adapted code, generated scripts) and are therefore also excluded from CRAP.
 */
export async function readBiomeLinterDisabledPaths(projectRoot: string): Promise<string[]> {
	const biomePath = resolve(projectRoot, "biome.jsonc");
	const file = Bun.file(biomePath);
	if (!(await file.exists())) return [];
	const raw = await file.text();
	// biome.jsonc allows comments — strip them before JSON.parse. We only
	// strip line comments; the file does not currently use block comments.
	const stripped = raw.replace(/^\s*\/\/.*$/gm, "");
	const config = JSON.parse(stripped) as {
		overrides?: { includes?: string[]; linter?: { enabled?: boolean } }[];
	};
	const out: string[] = [];
	for (const override of config.overrides ?? []) {
		if (override.linter?.enabled === false) {
			for (const p of override.includes ?? []) out.push(p);
		}
	}
	return out;
}

/**
 * Read `crap.ignore.json` and return its `ignore` array. This list decouples
 * the CRAP gate from lint enforcement — pure-UI React files (which we still
 * lint) can be excluded from CRAP scoring without disabling biome on them.
 * Returns an empty array if the file is missing.
 */
export async function readCrapIgnorePaths(projectRoot: string): Promise<string[]> {
	const ignorePath = resolve(projectRoot, "crap.ignore.json");
	const file = Bun.file(ignorePath);
	if (!(await file.exists())) return [];
	const raw = await file.text();
	const config = JSON.parse(raw) as { ignore?: string[] };
	return config.ignore ?? [];
}

/**
 * Convert a glob-ish pattern into an anchored RegExp. We handle:
 *   - `**`  → `.*`        (any number of path segments, including slashes)
 *   - `*`   → `[^/]*`     (anything within a single segment)
 *   - everything else escaped as a literal
 * The pattern is anchored with `^…$`.
 */
function globToRegExp(pattern: string): RegExp {
	let out = "";
	for (let i = 0; i < pattern.length; i++) {
		const ch = pattern[i];
		if (ch === "*") {
			if (pattern[i + 1] === "*") {
				out += ".*";
				i++;
			} else {
				out += "[^/]*";
			}
		} else if (ch !== undefined && /[.+?^${}()|[\]\\]/.test(ch)) {
			out += `\\${ch}`;
		} else {
			out += ch;
		}
	}
	return new RegExp(`^${out}$`);
}

function pathMatchesAnyPattern(path: string, patterns: string[]): boolean {
	for (const p of patterns) {
		if (p === path) return true;
		// Fast path: trailing /** matches any descendant.
		if (p.endsWith("/**") && !p.slice(0, -3).includes("*")) {
			const prefix = p.slice(0, -3);
			if (path.startsWith(`${prefix}/`)) return true;
			continue;
		}
		// General glob — `*` segments may appear in the middle, e.g.
		// `src/entities/*/ui/**` → `src/entities/audio-device/ui/Foo.tsx`.
		if (p.includes("*")) {
			if (globToRegExp(p).test(path)) return true;
		}
	}
	return false;
}

/**
 * Returns the set of 1-indexed line numbers whose nearest preceding non-blank
 * line carries a `@crap-exclude` annotation. We scan up to 5 lines preceding
 * the target line for the marker — enough to clear a typical comment block
 * and a single dependency-array closing bracket, but tight enough not to
 * leak across unrelated declarations.
 */
function collectCrapExcludeLines(text: string): Set<number> {
	const lines = text.split(/\r?\n/);
	const out = new Set<number>();
	for (let i = 0; i < lines.length; i++) {
		if (!lines[i]?.includes("@crap-exclude")) continue;
		// Mark this line and the next 6 lines (1-indexed) as exclusion-bearing.
		for (let j = 0; j <= 6; j++) out.add(i + 1 + j);
	}
	return out;
}

export async function analyze(
	projectRoot: string,
	coverage: LcovCoverage,
	options: AnalyzeOptions
): Promise<FunctionMetric[]> {
	const out: FunctionMetric[] = [];
	const sourceGlob = new Glob("**/*.{ts,tsx}");
	const excludePaths = options.excludePaths ?? [];

	for (const root of options.roots) {
		const cwd = resolve(projectRoot, root);
		for await (const rel of sourceGlob.scan({ cwd, absolute: false })) {
			if (!options.include.test(rel) || options.exclude.test(rel)) continue;
			const fullRel = `${root}/${rel}`.replace(/\\/g, "/");
			if (pathMatchesAnyPattern(fullRel, excludePaths)) continue;
			const text = await Bun.file(resolve(projectRoot, fullRel)).text();
			const excludedLines = collectCrapExcludeLines(text);
			const fns = parseFunctions(fullRel, text);
			const fileCoverage = coverage.get(fullRel) ?? coverage.get(fullRel.replace(/\//g, "\\"));
			for (const fn of fns) {
				if (excludedLines.has(fn.startLine)) continue;
				const cov = fileCoverage
					? functionCoverageFraction(fileCoverage.lineHits, fn.startLine, fn.endLine)
					: null;
				out.push({
					file: fullRel,
					name: fn.name,
					startLine: fn.startLine,
					endLine: fn.endLine,
					complexity: fn.complexity,
					coverage: cov,
					crap: calculateCrapScore(fn.complexity, cov),
				});
			}
		}
	}
	return out;
}

/**
 * Coverage fraction = hit lines / executable lines within the function span.
 * "Executable lines" are any lines mentioned in the LCOV DA: records that
 * fall within [startLine, endLine]. Lines outside the LCOV record set are
 * treated as non-executable (comments, declarations, type-only code).
 */
function functionCoverageFraction(
	lineHits: Map<number, number>,
	startLine: number,
	endLine: number
): number | null {
	let executable = 0;
	let hit = 0;
	for (const [line, hits] of lineHits) {
		if (line < startLine || line > endLine) continue;
		executable++;
		if (hits > 0) hit++;
	}
	if (executable === 0) return null;
	return hit / executable;
}
